### ISSUE TITLE
Node-link-neighbor data structures

### ISSUE THREAD

**Issue by gregtucker** [landlab/landlab#191]
What are our current data structures for storing:
1 - the links attached to each node, and
2 - the neighbors of each node?
One set I know about, because I created it, is: node_inlink_matrix and node_outlink_matrix. There seem to be others, but I'm struggling to figure out what they are exactly. I'd like to clean this up so we have just ONE consistent set of data structures with a reasonable balance between algorithm efficiency and storage.

Specifically, I want to explore the utility of replacing our current data structures with the following, which are meant to represent all grid types:

1 - links_at_node: NxM matrix containing IDs of links that connect to each node (M=4 for raster, 6 for hex, and MAX_NUM_NEIGHBORS for Voronoi)

2 - link_signs_at_node: NxM matrix containing 1 if node is the link's head, -1 if it's the tail, and 0 if there is no link at this position

3 - active_link_signs_at_node: NxM matrix containing 1 if node is the link's head, -1 if it's the tail, and 0 if there is no link at this position OR if link is inactive

4 - neighbors_at_node: NxM matrix containing IDs of neighbor nodes (corresponding exactly to links_at_node)

If, for example, you want to mask out inactive links or neighbors on inactive links, you mask out the zero elements of the active_link_signs_at_node array.

These, or any other, data structures must be compatible with (at least) the following operations:
1) flux divergence
2) steepest descent / flow routing
3) neighbor search/processing operations
4) mapping vector values at links back onto nodes (and vice versa)
5) D-infinity flow routing (slopes of triangles)

**Comment by SiccarPoint** [comment]
We already have a good subset of these - with the proviso that they are not returned in a compatible fashion, which is very, very annoying (see #156). The set is (for raster, at least...):

1. mg.node_links()  -> (4, nnodes), SWNE. Needs renaming.
2,3. We don't have these, yet.
4. mg.get_neighbor_list() -> (nnodes, 4), ENWS. Also needs renaming.

On the plus side, note that the need for renaming lets us keep these old functions a bit (thus not breaking the absolute ton of complex code I've written that already works around their indeosyncracies), while still introducing the new, correct forms.

I'd also rephrase your final paragraph a bit - we already do all these things, working with what we have, so clearly what we have already meets your criteria! The priorities for me are actually twofold: that doing these things going forward gets *easier*, but modifying what we have already doesn't become a complete nightmare.

**Comment by SiccarPoint** [comment]
So, immediate, non-controversial priorities:

1. Deprecate these functions, BUT LEAVE THEM THERE
2. Create (all as properties, I'd advocate):
    mg.links_at_node
    mg.link_signs_at_node (..."link_directions"?)
    mg.active_link_signs_at_node
    mg.neighbors_at_node

Things we already agreed on, according to my notes from last week:
* These are shape (nnodes, 4) (or 6, or nlinks, or whatever)
* Order is S,W,N,E
* We use -1s *in these cases* to indicate bad_index. This is important to let the calcs work smoothly. However, if we made them methods, not @properties, we could offer an argument to override this (I've done this in some of the older functions, and it works well).

@gregtucker: looks like you've progressed from the call last week, when you were advocating link_directions (direct equivalent of link_signs_at_node) and active_links_at_node, a BOOL. So, as I understand it, active_link_signs_at_node = link_directions * active_links_at_node. Clearly these are direct equivalents, but was wondering why you changed from this scheme?

**Comment by gregtucker** [comment]
I probably wasn't clear last week; this is basically the same scheme. No bools, just (-1, 0, 1)

**Comment by mcflugen** [comment]
This sounds great.
* Instead of `link_signs_at_node`, how about `link_dir_at_node`?
* Is there some way we can get rid of `active_link_signs_at_node`? Can both an inactive link and an absent link have `dir == 0`? If so, `active_link_signs_at_node` seems redundant.
* We should do this in stages, for sure, but I think we should get rid of the old functions as quickly as possible.
* I think we should revisit the ordering. It really would be nice if things were ordered consistently. I think it will be a pain rewriting things to be consistent, regardless of the direction.

An alternative for the `links_at_node` data structure could be a flat array called `links_at_node` (and `link_dir_at_node`) along with another flat array called `number_of_links_at_node`. It's not clear to me if it would play well with your use-cases or not, though. If we were writing this in C, I think it would be the way to go but, perhaps, it's not right for us.

**Comment by SiccarPoint** [comment]
Re a possible flat array, I don’t think I like this idea from an ease of use perspective. The way this works in CHILD is pretty opaque, and I don’t think it’s adequately user friendly, or really very pythonic (or indeed "Landlabbic”). Everywhere else the scheme is “give me the X at the Y”, and it comes back (max(len(X)), len(Y)) shaped. I think that’s super clear and a nice style to have. I anticipate it’ll be faster too, especially as Greg is basically proposing a mask. Then the “raggedness” for voronoi looks like (I’m not differentiating 1 and -1 here): 

    links_at_node = array([[ A, B, C, D,-1,-1,-1,-1],
                           [ E, F, G, H, I,-1,-1,-1],
                           [ J, K, L, M, N, O, -1, -1],
                           …])
    link_dir_at_node = array([[ 1, 1, 1, 1, 0, 0, 0, 0],
                              [ 1, 1, 1, 1, 1, 0, 0, 0],
                              [ 1, 1, 1, 1, 1, 1, 0, 0],
                              …])  # or whatever name we end up with

I really like how clean this is. The only downside I think will be disk space…?

**Comment by SiccarPoint** [comment]
(repeated comment with markdown for clarity) Re a possible flat array, I don’t think I like this idea from an ease of use perspective. The way this works in CHILD is pretty opaque, and I don’t think it’s adequately user friendly, or really very pythonic (or indeed "Landlabbic”). Everywhere else the scheme is “give me the X at the Y”, and it comes back (max(len(X)), len(Y)) shaped. I think that’s super clear and a nice style to have. I anticipate it’ll be faster too, especially as Greg is basically proposing a mask. Then the “raggedness” for voronoi looks like (I’m not differentiating 1 and -1 here): 

    links_at_node = array([[ A, B, C, D,-1,-1,-1,-1],
                           [ E, F, G, H, I,-1,-1,-1],
                           [ J, K, L, M, N, O, -1, -1],
                           …])
    link_dir_at_node = array([[ 1, 1, 1, 1, 0, 0, 0, 0],
                              [ 1, 1, 1, 1, 1, 0, 0, 0],
                              [ 1, 1, 1, 1, 1, 1, 0, 0],
                              …])  # or whatever name we end up with

I really like how clean this is. The only downside I think will be disk space…?

**Comment by gregtucker** [comment]
Actually, CHILD doesn't use flat arrays. The Braun and Willett algorithm *does*: it stores neighbors in a flat array, and retrieves them by use of indexing: you know that the neighbors of i are found at indices[i:i+1] in the array.

I like "dir" rather than "sign", thanks.

The point of active_link_dir_at_node is to save space. I actually don't see an obvious way around it. Suppose node i has links 1, 2, 3, and 4, with the first 2 coming in and the second two going out. But only 2 and 3 are active. You could store this info like so:

links_at_node[i] = [1, 2, 3, 4]
link_dirs_at_node[i] = [1, 1, -1, -1]
active_link_dirs_at_node[i] = [0, 1, -1, 0]

or you could store it like:

links_at_node[i] = [1, 2, 3, 4]
link_dirs_at_node[i] = [1, 1, -1, -1]
active_links_at_node[i] = [-1, 2, 3, -1]

But in the latter case I think you'd also still need
active_link_dirs_at_node[i] = [0, 1, -1, 0]

Part of the point of all this is to be able to calculate flux divergence:

net_flux[i] = sum(flux[links_at_node[i]]*link_dirs_at_node[i])
(which includes fluxes at all links: two incoming, two outgoing)

or

net_flux[i] = sum(flux[links_at_node[i]]*active_link_dirs_at_node[i])
(which includes only fluxes at active links; the others are zero)

Note that this could be vectorized (i.e., numpy'd).

A flat array would also work ok but I think would need to be Cythonized for speed; at any rate, I can't think of a way to "arrayify" such a calculation.

**Comment by mcflugen** [comment]
I was thinking that if a node was "inactive" maybe we could think of it not being there at all. So your example would be,
``` python
links_at_node[i] = [1, 2, 3, 4]
link_dirs_at_node[i] = [0, 1, -1, 0]
```
The `links_at_node` array indicates that there are, in fact, four links at node *i*, but the `link_dir_at_node` array says links 1 and 4 are inactive. You do lose the directionality of those two links, though. This could be a problem if you wanted to later activate them.

I agree, a flat-array solution would probably need Cythonizing. I think that might be it's biggest problem. Being able to use numpy arrays makes things much easier.

**Comment by gregtucker** [comment]
Yes, that would work, but then we would no longer have the option of doing fluxes at *all* links.

The flat array idea is intriguing... I'm tempted to try both with a couple of use cases and see how they compare.

**Comment by mcflugen** [comment]
I wasn't sure if we ever want to calculate fluxes using inactive links. If we do, then, for sure that wouldn't work.

I'm not certain about the flat arrays. I only brought it up so we could say we made a decision based on `n_choices > 1`.

**Comment by gregtucker** [comment]
I agree it would be unlikely... but there may be other reasons to want directional information for inactive links.

Decision making is always so much harder with n_choices > 1!

**Comment by jadams15** [comment]
@mcflugen - I thought we had agreed SWNE, and here is something from last week? Am I missing something?

**Comment by mcflugen** [comment]
@SiccarPoint, snuck that in on me. Since we agreed to angles being measured counter clockwise starting at the positive x-axis, I think, ultimately, it would be good to do all our ordering that way. It seems a little random to start ordering clockwise starting at the negative y-axis. I'm fine keeping ordering of neighbor nodes as SWNE for now but, if possible, would like to eventually see things move to ENWS. But... well... that's just like my opinion, man. I'm cool with whatever, but we do need to be consistent.
